src/main/java/ca/corefacility/bioinformatics/irida/model/sample/Sample.java

```diff
@@ -338,27 +338,30 @@ public void setMetadata(Map<MetadataTemplateField, MetadataEntry> inputMetadata)
 	}
 	
 	/**
-	 * Merge {@link MetadataEntry} into the sample's existing metadata collection. Duplicate
-	 * keys will be overwritten.
+	 * Merge {@link MetadataEntry} into the sample's existing metadata collection.
+	 * Duplicate keys will be overwritten.
 	 * 
-	 * @param inputMetadata
-	 *            the metadata to merge into the sample
+	 * @param inputMetadata the metadata to merge into the sample
 	 */
 	public void mergeMetadata(Map<MetadataTemplateField, MetadataEntry> inputMetadata) {
 		// loop through entry set and see if it already exists
 		for (Entry<MetadataTemplateField, MetadataEntry> entry : inputMetadata.entrySet()) {
+			MetadataEntry newMetadataEntry = entry.getValue();
 
-			// if the value isn't empty
-			if (!entry.getValue().getValue().isEmpty()) {
+			// if the key is found, replace the entry
+			if (metadata.containsKey(entry.getKey())) {
+				MetadataEntry originalMetadataEntry = metadata.get(entry.getKey());
 
-				// if the key is found, replace the entry
-				if (metadata.containsKey(entry.getKey())) {
-					MetadataEntry metadataEntry = metadata.get(entry.getKey());
-					metadataEntry.setValue(entry.getValue().getValue());
+				// if the metadata entries are of the same type, I can directly merge
+				if (originalMetadataEntry.getClass().equals(newMetadataEntry.getClass())) {
+					originalMetadataEntry.merge(newMetadataEntry);
 				} else {
-					// otherwise add the new entry
-					metadata.put(entry.getKey(), entry.getValue());
+					// if they are different types, I need to replace the metadata entry instead of merging
+					metadata.put(entry.getKey(), newMetadataEntry);
```

Comment:
Let me understand this section a bit.  This is testing for a case where someone makes a metadata entry named "bob", then later a pipeline comes along and writes an entry named "bob".  Since the pipeline entry should be a different type, it needs to create a new entry.  Is that correct?

Reply:
Yes, correct. If we did not create a new `PipelineProvidedMetadataEntry` in this case then the associated `AnalysisSubmission` would be lost when the metadata is updated (since `MetadataEntry` does not store this information).

Reply:
One question I had was whether or not I should delete the old `MetadataEntry` in this case, since now it's unassociated with a sample. I don't think a `MetadataEntry` can be associated with multiple samples right now.  Deleting would likely require moving all this code to a service method though and/or playing around with delete cascade functionality.

Reply:
At the moment a `MetadataEntry` is only associated with one sample, but that's probably going to change as we get into the onto stuff.  I've been thinking about a good way to handle this, whether we should have a job that runs to garbage collect dead entries or something.  I'm not sure I have a good answer to that question yet.

Reply:
No problem. It's not going to hurt anything to have some unassociated `MetadataEntry`s in our database. And this particular situation is unlikely to come up right now anyways (unless someone uploads metadata with same field names to a project before running pipelines).